internal/controllers/import_controller_v3.go

```diff
@@ -0,0 +1,393 @@
+
+	if r.remoteClientGetter == nil {
+		r.remoteClientGetter = remote.NewClusterClient
+	}
+
+	capiPredicates := predicates.All(log,
+		predicates.ResourceHasFilterLabel(log, r.WatchFilterValue),
+		turtlespredicates.ClusterWithoutImportedAnnotation(log),
+		turtlespredicates.ClusterWithReadyControlPlane(log),
+		turtlespredicates.ClusterOrNamespaceWithImportLabel(ctx, log, r.Client, importLabelName),
+	)
+
+	c, err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterv1.Cluster{}).
+		WithOptions(options).
+		WithEventFilter(capiPredicates).
+		Build(r)
+	if err != nil {
+		return fmt.Errorf("creating new controller: %w", err)
+	}
+
+	// Watch Rancher managementv3 clusters
+	err = c.Watch(
+		source.Kind(mgr.GetCache(), &managementv3.Cluster{}),
+		handler.EnqueueRequestsFromMapFunc(r.rancherClusterToCapiCluster(ctx, capiPredicates)),
+	)
+	if err != nil {
+		return fmt.Errorf("adding watch for Rancher cluster: %w", err)
+	}
+
+	ns := &corev1.Namespace{}
+	err = c.Watch(
+		source.Kind(mgr.GetCache(), ns),
+		handler.EnqueueRequestsFromMapFunc(namespaceToCapiClusters(ctx, capiPredicates, r.Client)),
+	)
+
+	if err != nil {
+		return fmt.Errorf("adding watch for namespaces: %w", err)
+	}
+
+	r.recorder = mgr.GetEventRecorderFor("rancher-turtles")
+	r.controller = c
+	r.externalTracker = external.ObjectTracker{
+		Controller: c,
+	}
+
+	return nil
+}
+
+// +kubebuilder:rbac:groups="",resources=secrets;events;configmaps,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=coordination.k8s.io,resources=leases,verbs=get;create;update
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;clusters/status,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=*,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=management.cattle.io,resources=clusters;clusters/status,verbs=get;list;watch;create;update;delete;deletecollection;patch
+// +kubebuilder:rbac:groups=management.cattle.io,resources=clusters;clusterregistrationtokens;clusterregistrationtokens/status,verbs=get;list;watch
+
+// Reconcile reconciles a CAPI cluster, creating a Rancher cluster if needed and applying the import manifests.
+func (r *CAPIImportManagementV3Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, reterr error) {
+	log := log.FromContext(ctx)
+	log.Info("Reconciling CAPI cluster")
+
+	capiCluster := &clusterv1.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, capiCluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	if capiCluster.ObjectMeta.DeletionTimestamp.IsZero() && !turtlesannotations.HasClusterImportAnnotation(capiCluster) {
+		if !controllerutil.ContainsFinalizer(capiCluster, managementv3.CapiClusterFinalizer) {
```

Comment:
```suggestion
	if capiCluster.ObjectMeta.DeletionTimestamp.IsZero() && !turtlesannotations.HasClusterImportAnnotation(capiCluster) && !controllerutil.ContainsFinalizer(capiCluster, managementv3.CapiClusterFinalizer) {
```
But to be honest this feels like it belongs under reconcileNormal scenario

Reply:
The reason I opted for having this here is to avoid problems due to changes to the object. Since the capi cluster is being updated by the capi controller and then we're also applying the finalizer, doing it in `reconcileNormal` would mean that we would have to deal with the changed object and then repeat the same when patching after reconcile [here](https://github.com/rancher-sandbox/rancher-turtles/blob/e1ab73b9b19078b10c4fd3c23496ee51e438724f/internal/controllers/import_controller_v3.go#L174). I think the downside of having it here is that there's a corner case in which the finalizer is set on the capi cluster and then the rancher cluster fails to create, but I'd say this is better than the situation in which we have a rancher cluster created but we fail to add the finalizer to the capi cluster.